### PR TITLE
Improve the way we init our RPC servers 

### LINKED ISSUE #1319: dev: Maybe we can improve the way we init our RPC servers
In `crates/node/src/rpc/mod.rs` we init one `Starknet` instance for each one of the RPC server extensions we have.
Maybe there is some opti to do there.
Either share a single `Starknet` instance for all, inside an Arc. Or something else.
Also, not every RPC server we define needs all of the fields inside `Starknet`.

Something to think about, but not crucial performance-wise.

### PR DESCRIPTION
**Pull request by hhamud**
# Pull Request type
* Fixes #1319

Please add the labels corresponding to the type of changes your PR introduces:
- Feature
- Refactoring (no functional changes, no API changes)

## What is the current behavior?
The issue here was that multiple starknet instances were being needlessly created as mentioned here #1319 

Resolves: #NA

## What is the new behavior?
It refactors the `create_full` function in `crates/node/src/rpc/mod.rs` by creating an Arc instance of starknet struct by creating a new wrapper type `Astarknet`.  In `crates/client/rpc/src/lib.rs

The traits were re-applied to `Astarknet` as Arc is not a fundamental type and therefore would violate the orphan rule. 

## Does this introduce a breaking change?
No

### CODE REVIEW

File: crates/client/rpc/src/lib.rs [hunk 1]
```diff
@@ -1868,3 +1873,599 @@ fn actual_resources_to_execution_resources(resources: ResourcesMapping) -> Execu
         keccak_builtin_applications: *resources.get("keccak_builtin").unwrap_or(&0),
     }
 }
+
+// Newtype Wrapper to escape Arc orphan rules
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:1877]: Can you just put all this in a separate module, this file is already too long :)

**Review by tdelabro — CHANGES_REQUESTED** [review]
Damn, good job!
Especially for a first PR!

**Comment by tdelabro** [comment]
@hhamud add a line in the changelog explaining what you do in this PR, it will unlock the CI

**Comment by tdelabro** [comment]
@hhamud lintrs are failing now, run:
```shell
 $ npx prettier -w .
 $  npx @taplo/cli fmt --config ./taplo/taplo.toml
 ```
 and cargo fmt and cargo clippy

**Comment by hhamud** [comment]
> @hhamud lintrs are failing now, run:
> 
> ```shell
>  $ npx prettier -w .
>  $  npx @taplo/cli fmt --config ./taplo/taplo.toml
> ```
> 
> and cargo fmt and cargo clippy

It might be worth having a git hook requiring those before making a push. Its very easy to forget to do those.

**Comment by tdelabro** [comment]
@hhamud It's still failing but it's not your fault. I'm pushing a fix to main and will rebase and merge your PR after that

**Comment by tdelabro** [comment]
`thread 'tokio-runtime-worker' has overflowed its stack`
https://github.com/keep-starknet-strange/madara/actions/runs/8007193998/job/21871319531?pr=1462

It looks like it's causing problems with Tokio...
I'm not sure why

@hhamud

**Comment by hhamud** [comment]
> `thread 'tokio-runtime-worker' has overflowed its stack` https://github.com/keep-starknet-strange/madara/actions/runs/8007193998/job/21871319531?pr=1462
> 
> It looks like it's causing problems with Tokio... I'm not sure why
> 
> @hhamud

I'll investigate why. Do I need anything specific to run these specific tests locally?

**Comment by tdelabro** [comment]
@hhamud to run locally:
```shell
$  cargo run -- --dev
$ cd  starknet-rpc-test && cargo test
```
In two different terms, look at the logs in the first one

**Comment by hhamud** [comment]
I am not sure why but when I try to run the toml linter it fails
```shell
user@192 madara % npx @taplo/cli fmt --config ./taplo/taplo.toml
 INFO taplo:format_files:load_config: found configuration file path="./taplo/taplo.toml"
 INFO taplo:format_files:collect_files: found files total=44 excluded=0 cwd="/Users/user/Documents/projects/madara"
panicked at /Users/JounQin/Workspaces/GitHub/taplo/crates/taplo/src/formatter/mod.rs:684:25:
assertion failed: entry.comment.is_none()

Stack:

Error
    at I.wbg.__wbg_new_abda76e883ba8a5f (/Users/user/.npm/_npx/29347d57df19087c/node_modules/@taplo/cli/dist/cli.js:42:4218)
    at console_error_panic_hook::Error::new::hd83448d5e1a1b88a (<anonymous>:wasm-function[40242]:0x1249bc2)
    at console_error_panic_hook::hook_impl::h24434d5d3a31a46a (<anonymous>:wasm-function[10411]:0xcf17fe)
    at console_error_panic_hook::hook::hb562c91cafc2c895 (<anonymous>:wasm-function[59765]:0x12e59c7)
    at core::ops::function::Fn::call::h46d6beedf89bee9d (<anonymous>:wasm-function[42879]:0x1275769)
    at std::panicking::rust_panic_with_hook::h8e8556df5e46d2ff (<anonymous>:wasm-function[20551]:0xfc8712)
    at std::panicking::begin_panic_handler::{{closure}}::h142b9f459ded72d9 (<anonymous>:wasm-function[25954]:0x10c54b9)
    at std::sys_common::backtrace::__rust_end_short_backtrace::h1c0d3278570375be (<anonymous>:wasm-function[61550]:0x12e9f20)
    at rust_begin_unwind (<anonymous>:wasm-function[42801]:0x12744d7)
    at core::panicking::panic_fmt::hbea4287e435ed8a3 (<anonymous>:wasm-function[43645]:0x1280e86)


wasm://wasm/065ed00a:1


RuntimeError: unreachable
    at __rust_start_panic (<anonymous>:wasm-function[63493]:0x12ec0f3)
    at rust_panic (<anonymous>:wasm-function[56261]:0x12dbf09)
    at std::panicking::rust_panic_with_hook::h8e8556df5e46d2ff (<anonymous>:wasm-function[20551]:0xfc8742)
    at std::panicking::begin_panic_handler::{{closure}}::h142b9f459ded72d9 (<anonymous>:wasm-function[25954]:0x10c54b9)
    at std::sys_common::backtrace::__rust_end_short_backtrace::h1c0d3278570375be (<anonymous>:wasm-function[61550]:0x12e9f20)
    at rust_begin_unwind (<anonymous>:wasm-function[42801]:0x12744d7)
    at core::panicking::panic_fmt::hbea4287e435ed8a3 (<anonymous>:wasm-function[43645]:0x1280e86)
    at core::panicking::panic::haa9f794b7d1c74c9 (<anonymous>:wasm-function[41361]:0x125d8f9)
    at taplo::formatter::add_entries::hed6838edc1b01c33 (<anonymous>:wasm-function[437]:0x3e5c50)
    at taplo::formatter::format_root::hcac06d9d442d0aa3 (<anonymous>:wasm-function[256]:0x2eb282)
```

File: crates/client/rpc/src/astarknet.rs [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+    ///
+    /// # Returns
+    ///
+    /// * `fee_estimate` - fee estimate in gwei
+    async fn estimate_fee(
+        &self,
+        request: Vec<BroadcastedTransaction>,
+        block_id: BlockId,
+    ) -> RpcResult<Vec<FeeEstimate>> {
+        let substrate_block_hash = self.0.substrate_block_hash_from_starknet_block(block_id).map_err(|e| {
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/rpc/src/astarknet.rs:357]: Which other function?

[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/client/rpc/src/astarknet.rs:357]: You should probably try the `<T as Trait> ` syntax to select the function you want

**Comment by tdelabro** [comment]
@hhamud about the toml linter I have the same issue, it's new.
Try installing taplo through cargo: https://taplo.tamasfe.dev/cli/installation/cargo.html
and then run 
```shell
 taplo fmt --config ./taplo/taplo.toml
```

It worked for me

File: crates/node/src/rpc/mod.rs [hunk 3]
```diff
@@ -70,46 +71,24 @@ where
...
-        starknet_params.madara_backend,
-        starknet_params.overrides,
-        pool,
-        graph,
-        starknet_params.sync_service,
-        starknet_params.starting_block,
-        starknet_params.genesis_provider,
-    )))?;
+
+    // Arc<Starknet
```
Comment (tdelabro) [inline crates/node/src/rpc/mod.rs:75]: Remove this comment

File: crates/client/rpc/src/lib.rs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 //!
 //! It uses the madara client and backend in order to answer queries.
 
+pub mod astarknet;
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:5]: Can we think of a better name than `astrarknet` I'm not sure of what you meant by that.
Can we name it `StarknetRpcWrapper` or something along those lines?

**Review by tdelabro — APPROVED** [review]

File: crates/client/rpc/src/lib.rs (hunk 4 above)
Comment (hhamud) [reply crates/client/rpc/src/lib.rs:5]: I called it Astarknet for short of arcStarknet but I can just change it to StarknetRpcWrapper. Do you want to change the name of the file, the mod and every instance of the name?

Comment (tdelabro) [reply crates/client/rpc/src/lib.rs]: `StarknetRpcWrapper` is not a really good name either haha

**Pull request merged**